### PR TITLE
GH-40878: [JAVA] Fix flight-sql-jdbc-driver shading issues

### PR DESCRIPTION
**Pull request by laurentgo**
### Rationale for this change

The `flight-sql-jdbc-driver` jar is not shaded properly:
* a reduced pom.xml file is not generated. The published pom.xml file declares dependencies which are actually present in the jar and should not be fetched externally
* several classes/files are not relocated properly

### What changes are included in this PR?

Fix pom.xml and relocations. Also removes annotations dependencies and include a integration test to prevent future breakage.

### Are these changes tested?

Yes. A new integration test check the jar content

### Are there any user-facing changes?

Yes. The published pom.xml file on Maven will be cleaned of any dependency
* GitHub Issue: #40878

### CODE REVIEW

**Review by jbonofre — APPROVED** [review]
LGTM. Shading simplifies driver use :)

File: java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import org.junit.rules.ErrorCollector;
+import org.junit.rules.TestRule;
+import org.junit.rules.Timeout;
+
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Check the content of the JDBC driver jar
+ *
+ * After shading everything should be under com.dremio.jdbc package
```
Comment (vibhatha) [inline java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java:43]: Appreciate a bit of explanation on the comment `com.dremio.jdbc`?

Comment (jbonofre) [reply java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java:43]: That's a copy of an internal fix. Let me fix. Sorry about that.

Comment (jbonofre) [inline java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java:43]: It should be:

```
After shading everything should be under cfjd package
```

**Review by jbonofre — CHANGES_REQUESTED** [review]

File: java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java (hunk 1 above)
Comment (laurentgo) [reply java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java:43]: My bad. I was working simultaneously on a fix for https://github.com/dremio/dremio-oss/blob/master/distribution/jdbc-driver/src/test/java/com/dremio/jdbc/ITValidateShadedJar.java and mixed up the comments by accident. Will fix asap

Comment (jbonofre) [reply java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java:43]: No problem. Thanks !

NB: I tried to push a commit fixing that on your branch, but I don't have permission to do so. I let you fix that.

Comment (laurentgo) [reply java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java]: Fix pushed

**Review by jbonofre — APPROVED** [review]

**Review by lidavidm — APPROVED** [review]
Thank you!

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

File: java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java (hunk 1 above)
Comment (laurentgo) [reply java/flight/flight-sql-jdbc-driver/src/test/java/org/apache/arrow/driver/jdbc/ITDriverJarValidation.java]: Sorry, I was working for a fix for https://github.com/dremio/dremio-oss/blob/master/distribution/jdbc-driver/src/test/java/com/dremio/jdbc/ITValidateShadedJar.java at the same time, and I mixed up the comments.

I uploaded a change to the comment. Issue should be fixed

**Pull request merged**